k4EDM4hep2LcioConv/src/k4Lcio2EDM4hepConv.cpp

```diff
@@ -0,0 +1,991 @@
+    const TypeMapT<const lcio::Cluster*, edm4hep::MutableCluster>& clusterMap,
+    const TypeMapT<const lcio::Track*, edm4hep::MutableTrack>& tracksMap)
+  {
+    int edmnum = 1;
+    for (auto& [lcio, edm] : recoparticlesMap) {
+      edmnum++;
+
+      auto vertex = lcio->getStartVertex();
+      if (const auto it = vertexMap.find(vertex); it != vertexMap.end()) {
+        edm.setStartVertex(it->second);
+      }
+      else {
+        std::cerr << "Cannot find corresponding EDM4hep Vertex for a LCIO Vertex, "
+                     "while trying to resolve the ReconstructedParticle Relations"
+                  << std::endl;
+      }
+
+      auto clusters = lcio->getClusters();
+      for (auto c : clusters) {
+        const auto it = clusterMap.find(c);
+        if (it != clusterMap.end()) {
+          edm.addToClusters(it->second);
+        }
+        else {
+          std::cerr << "Cannot find corresponding EDM4hep Cluster for a LCIO Cluster, "
+                       "while trying to resolve the ReconstructedParticle Relations"
+                    << std::endl;
+        }
+      }
+
+      auto tracks = lcio->getTracks();
+      for (auto t : tracks) {
+        const auto it = tracksMap.find(t);
+        if (it != tracksMap.end()) {
+          edm.addToTracks(it->second);
+        }
+        else {
+          std::cerr << "Cannot find corresponding EDM4hep Tracks for a LCIO Tracks, "
+                       "while trying to resolve the ReconstructedParticle Relations"
+                    << std::endl;
+        }
+      }
+
+      auto parents = lcio->getParticles();
+      for (auto p : parents) {
+        const auto it = recoparticlesMap.find(p);
+        if (it != recoparticlesMap.end()) {
+          edm.addToParticles(it->second);
+        }
+        else {
+          std::cerr << "Cannot find corresponding EDM4hep RecoParticle for a LCIO RecoParticle, "
+                       "while trying to resolve the ReconstructedParticles parents Relations"
+                    << std::endl;
+        }
+      }
+    }
+  }
+
+  void resolveRelationsSimCalorimeterHit(
+    TypeMapT<const lcio::SimCalorimeterHit*, edm4hep::MutableSimCalorimeterHit>& SimCaloHitMap,
+    const TypeMapT<const lcio::MCParticle*, edm4hep::MutableMCParticle>& mcparticlesMap)
+  {
+    // TODO: Currently not doing anything here because we cannot get a mutable
+    // CaloHitContribution from the SimCalorimeterHit via the podio generated
+    // interface. The underlying issue is https://github.com/AIDASoft/podio/issues/347
+
+    // for (auto& [lcio, edm] : SimCaloHitMap) {
+    //   auto contributionLen = lcio->getNMCParticles();
+    //   for (auto i = 0; i < contributionLen; i++) {
+    //     auto mcp = lcio->getParticleCont(i);
+    //     const auto it = mcparticlesMap.find(mcp);
+    //     if (it != mcparticlesMap.end()) {
+    //       edm.getContributions(i).setParticle(it->second); // This breaks
+    //     }
+    //   }
+    // }
```

Comment:
A possible way to work around this issue here would be to not create the `CaloHitContribution` collection in the `convertSimCalorimeterHit` method, but instead create it here and return it from here. In that case it would still be possible to set the particles to the `CaloHitContribution`. That approach would entail a bit of interface changes in some of the calling methods, as there would need to be a way to "transport" back the new collection to where it would need to go for a full event conversion.